### ISSUE TITLE
[Process] Some tests fail on Windows

### ISSUE THREAD

**Issue by maryo** [symfony/symfony#22556]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | master

Some tests of the process component fail on my machine - x64 Windows 10, PHP 7.1.4

```
D:\Www\symfony>php ./phpunit src/Symfony/Component/Process/
#!/usr/bin/env php
INFO: Could not find files for the given pattern(s).
PHPUnit 6.0.13 by Sebastian Bergmann and contributors.

Testing src/Symfony/Component/Process/
...SS.S.FF..................S...S..............................  63 / 161 ( 39%)
......SSS..S..............SSSS............SS..........SS....... 126 / 161 ( 78%)
....................E..............                             161 / 161 (100%)

Time: 23.04 seconds, Memory: 4.00MB

There was 1 error:

1) Symfony\Component\Process\Tests\ProcessTest::testInheritEnvDisabled
array_intersect_key(): Argument #1 is not an array

D:\Www\symfony\src\Symfony\Component\Process\Tests\ProcessTest.php:1443
D:\Www\symfony\.phpunit\phpunit-6.0\phpunit:5

--

There were 2 failures:

1) Symfony\Component\Process\Tests\PhpProcessTest::testNonBlockingWorks
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'hello world!'
+''

D:\Www\symfony\src\Symfony\Component\Process\Tests\PhpProcessTest.php:28
D:\Www\symfony\.phpunit\phpunit-6.0\phpunit:5

2) Symfony\Component\Process\Tests\PhpProcessTest::testCommandLine
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-7.1.4cli
+

D:\Www\symfony\src\Symfony\Component\Process\Tests\PhpProcessTest.php:46
D:\Www\symfony\.phpunit\phpunit-6.0\phpunit:5

ERRORS!
Tests: 161, Assertions: 252, Errors: 1, Failures: 2, Skipped: 17.

Legacy deprecation notices (12)
```

**Comment by nicolas-grekas** [comment]
I can't reproduce on appveyor, see https://ci.appveyor.com/project/fabpot/symfony/build/1.0.21543: this doesn't fail for the Process component. Are these errors predictable on your machine? Can you please try to investigate a bit further where this might come from?

**Comment by maryo** [comment]
@nicolas-grekas 
The main problem was #22613. And I don't know why but the first errored test was still causing an error `1) Symfony\Component\Process\Tests\ProcessTest::testInheritEnvDisabled
array_intersect_key(): Argument #1 is not an array`
So I was looking into why it happens and then somehow the error magically disappeared. But I encountered another issue described in #22614 so maybe it's related

**Comment by maryo** [comment]
@nicolas-grekas 
I've found sime time to look at it again primarily because of #22614 (I've updated the PR to include a fix).
I've run the tests on my home PC (also Windows 10 x64, PHP 7.1 x64) and even more of them failed :-/.

```
1) Symfony\Component\Process\Tests\ProcessTest::testIteratorInput
Symfony\Component\Process\Exception\ProcessTimedOutException: The process ""D:\Program Files\PHP\php.exe" -r "stream_copy_to_stream(STDIN, STDOUT);"" exceeded the timeout of 60 seconds.

D:\www\symfony\src\Symfony\Component\Process\Process.php:1332
D:\www\symfony\src\Symfony\Component\Process\Process.php:421
D:\www\symfony\src\Symfony\Component\Process\Process.php:216
D:\www\symfony\src\Symfony\Component\Process\Tests\ProcessTest.php:1214
D:\www\symfony\.phpunit\phpunit-6.0\phpunit:5

2) Symfony\Component\Process\Tests\ProcessTest::testSimpleInputStream
Symfony\Component\Process\Exception\ProcessTimedOutException: The process ""D:\Program Files\PHP\php.exe" -r "echo 'ping'; stream_copy_to_stream(STDIN, STDOUT);"" exceeded the timeout of 60 seconds.

D:\www\symfony\src\Symfony\Component\Process\Process.php:1332
D:\www\symfony\src\Symfony\Component\Process\Process.php:421
D:\www\symfony\src\Symfony\Component\Process\Tests\ProcessTest.php:1234
D:\www\symfony\.phpunit\phpunit-6.0\phpunit:5

3) Symfony\Component\Process\Tests\ProcessTest::testInputStreamWithGenerator
Symfony\Component\Process\Exception\ProcessTimedOutException: The process ""D:\Program Files\PHP\php.exe" -r "stream_copy_to_stream(STDIN, STDOUT);"" exceeded the timeout of 60 seconds.

D:\www\symfony\src\Symfony\Component\Process\Process.php:1332
D:\www\symfony\src\Symfony\Component\Process\Process.php:421
D:\www\symfony\src\Symfony\Component\Process\Tests\ProcessTest.php:1278
D:\www\symfony\.phpunit\phpunit-6.0\phpunit:5

--

There was 1 failure:

1) Symfony\Component\Process\Tests\ProcessTest::testInputStreamWithCallable
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-123
+1

D:\www\symfony\src\Symfony\Component\Process\Tests\ProcessTest.php:1263
D:\www\symfony\.phpunit\phpunit-6.0\phpunit:5

ERRORS!
Tests: 135, Assertions: 217, Errors: 3, Failures: 1, Skipped: 14.

Legacy deprecation notices (17)
```

All of them pass when I comment the whole condition here https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Process/Pipes/AbstractPipes.php#L123
But I guess that's not a solution. I spent a lot of time trying to fix it but I was not successful. I don't know how exactly `stream_select` works. I tried to call it only once before writing to standard input (It's called even when there's nothing to write) which fixed all except `testInputStreamWithCallable`.

Also `1) Symfony\Component\Process\Tests\ProcessTest::testInheritEnvDisabled
array_intersect_key(): Argument #1 is not an array` failed many times but then it somehow started to work without any modification.

PhpProcessTest was also failing but I already fixed it inside #22613 as I mentioned.

**Comment by ChadSikorra** [comment]
@maryo Is that the correct spot in AbstractPipes you are commenting? Doing that seems to make no difference on my machine. I can confirm the same test failures for me in PhpProcess on Windows 10 with PHP 7.1.

I can get mine to work slightly further by modifying Process to use powershell.exe instead of cmd.exe. This results in slightly garbled output (probably need to futz with powershell params). Another thing that makes it pass is setting bypass_shell to false and then adjusting the quotes for the command (Though this causes all sorts of other failures elsewhere and obviously isn't a solution, lack of PID, etc). Though that may help provide hints about what's going wrong.

**Comment by nicolas-grekas** [comment]
I'm closing because this is transient, thus impossible to fix.
@maryo of course if you can debug that and find the root cause, that'd be awesome.
